### PR TITLE
Adds check for already opened socket

### PR DESCRIPTION
**Pull request by VeijoPesonen**
### Description
Checks if the socket is already opened on the modem. 

### Pull request type
    [X] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Breaking change

### CODE REVIEW

**Comment by SeppoTakalo** [comment]
This worries me.
When can we end up in a situation where we already have socket connected, but application has lost the track of it?

Also, is there any verification that "ALREADY CONNECTED" means that it is already connected to this exact address we want it to be connected?

File: ESP8266/ESP8266.cpp [hunk 1]
```diff
@@ -404,26 +406,31 @@ nsapi_error_t ESP8266::open_udp(int id, const char* addr, int port, int local_po
...
-               && _parser.recv("OK\n");
+        done = _parser.send("AT+CIPSTART=%d,\"%s\",\"%s\",%d", id, type, addr, port);
     }
 
     if (done) {
+        if (!_parser.recv("OK\n")) {
+            if (_sock_already) {
+                _sock_already = false; // To be raised again by OOB msg
+                _smutex.unlock();
+                return NSAPI_ERROR_IS_CONNECTED;
```
Comment (SeppoTakalo) [inline ESP8266/ESP8266.cpp]: Is this correct?

If Socket returns `ERROR_IS_CONNECTED` it means that the socket is now connected, and you can start sending to it.

But this is not the case here.

**Comment by SeppoTakalo** [comment]
After reading the AT Instruction set, I'm still puzzled on what actually this "ALREADY CONNECTED" means.
Does it mean that this socket ID is already taken, or does it mean that same address, port, etc is already connected by this same ID, or other ID.

Whatever the case is, seems that we think the socket ID is not used, and ESP thinks it is. So we need to clean up. And I don't see any other clean up than we should close the existing socket ID that we don't know anything about.

**Comment by marcuschangarm** [comment]
We would probably get the best reliability if we did a hard reset using the reset pin and signaled to all sockets that connection had been lost. Do we have API for reconnecting to an access point?

**Comment by SeppoTakalo** [comment]
Maybe reset would be a bit overkill, if we know that we are still connected.
But maybe we should send AT+CIPCLOSE to the socket, and re-open with new parameters.

**Comment by marcuschangarm** [comment]
I'm fine with overkill, but your suggestion makes more sense! 😄

**Comment by VeijoPesonen** [comment]
@SeppoTakalo, @marcuschangarm please check if the latest commit addresses the issue.

**Comment by kjbracey-arm** [comment]
I'd kind of like a little debug tracing here - just to note that it's happened.

If we get here, it's because we already ascertained from our structures that `id` is free, right, so no need to do any further local state cleanup?

**Comment by kjbracey-arm** [comment]
Oh, and drop the recursion. Compiler will probably tailcall on a release or develop build, but won't in debug.

Make it a `goto` to the top of the function (cf `AtCmdParser` main loop).

File: ESP8266/ESP8266.cpp [hunk 2]
```diff
@@ -404,26 +406,36 @@ nsapi_error_t ESP8266::open_udp(int id, const char* addr, int port, int local_po
...
     if (done) {
+        if (!_parser.recv("OK\n")) {
+            if (_sock_already) {
+                _sock_already = false; // To be raised again by OOB msg
+                done = close(id);
+                if (!done) { // Shouldn't happen
+                    goto EXIT;
+                }
+                _smutex.unlock();
+                return open_udp(id, addr, port, local_port); // Retry
```
Comment (SeppoTakalo) [inline ESP8266/ESP8266.cpp]: Unlimited recursion here.

Either do a loop, or add parameter to this local function which tell whether you are allowed to retry.

Comment (SeppoTakalo) [reply ESP8266/ESP8266.cpp]: For example `open_udp(int id, const char* addr, int port, int local_port, bool allow_retry);

Then first call with `true`, then on second call with `false`

Or use some other logic, like `goto RETRY` with internal variable.
Unlimited retrying not allowed either.

File: ESP8266/ESP8266.cpp [hunk 3]
```diff
@@ -404,26 +406,36 @@ nsapi_error_t ESP8266::open_udp(int id, const char* addr, int port, int local_po
...
+        done = _parser.send("AT+CIPSTART=%d,\"%s\",\"%s\",%d", id, type, addr, port);
     }
 
     if (done) {
+        if (!_parser.recv("OK\n")) {
+            if (_sock_already) {
+                _sock_already = false; // To be raised again by OOB msg
+                done = close(id);
+                if (!done) { // Shouldn't happen
+                    goto EXIT;
```
Comment (SeppoTakalo) [inline ESP8266/ESP8266.cpp]: Should not happend --> Should not be tolerated.
Maybe `MBED_ERROR(...)` that causes verbal error message and stops the device.

File: ESP8266/ESP8266.cpp (hunk 2 above)
Comment (SeppoTakalo) [reply ESP8266/ESP8266.cpp]: Also, on the second loop, this should cause `MBED_ERROR` because we did not fix the issue by closing the socket.

**Comment by SeppoTakalo** [comment]
The logic is now much better. Just small changes to the recursion mechanism.

**Comment by VeijoPesonen** [comment]
@SeppoTakalo please check

**Comment by VeijoPesonen** [comment]
Will do, but I create a separate PR

> ERROR
> If the TCP connection is already established, the response is:
> ALREADY CONNECTED

**Pull request merged**